test/unit/ext/test_uv_pool.c

```diff
@@ -0,0 +1,132 @@
+#include "../../../src/lib/threadpool.h"
+#include "../../lib/runner.h"
+#include "../../lib/uv.h"
+#include "src/utils.h"
+
+TEST_MODULE(ext_uv_pool);
+
+/******************************************************************************
+ *
+ * threadpool
+ *
+ ******************************************************************************/
+
+enum { WORK_ITEMS_NR = 50 };
+
+struct fixture
+{
+	xx_loop_t xx_loop;
+	xx_work_t work_req;
+};
+
+static void loop_setup(struct fixture *f)
+{
+	int rc;
+
+	rc = uv_loop_init(&f->xx_loop.loop);
+	munit_assert_int(rc, ==, 0);
+
+	rc = xx_loop_init(&f->xx_loop);
+	munit_assert_int(rc, ==, 0);
+}
+
+static void bottom_work_cb(xx_work_t *req)
+{
+	munit_logf(MUNIT_LOG_INFO, "bottom_work_cb() req=%p id=%u widx=%u tid=%d",
+		   req,
+		   xx__thread_id(),
+		   req->work_req.thread_idx,
+		   gettid());
+
+	munit_assert_int(req->work_req.thread_idx, ==, xx__thread_id());
+}
+
+static void bottom_after_work_cb(xx_work_t *req, int status UNUSED)
+{
+	static int count = 0;
+	munit_logf(MUNIT_LOG_INFO,
+		   "bottom_after_work_cb() tid=%d req=%p count=%d", gettid(),
+		   req, count);
+
+	/*
+	 * Note: Close the uv_loop. The alternative can be seen in libuv v1.46
+	 *       around special handling
+	 *
+	 *      - for initialization:
+	 *      uv__handle_unref(&loop->wq_async);
+	 *      loop->wq_async.flags |= UV_HANDLE_INTERNAL;
+	 *
+	 *      - and for finalization:
+	 *      uv_walk() ... if (h->flags & UV_HANDLE_INTERNAL) continue;
+	 */
+	if (count == WORK_ITEMS_NR)
+		xx_loop_async_close(req->loop);
```

Comment:
How will we arrange for xx_loop_async_close to be called at the appropriate time in the real dqlite?

Reply:
I wanted to call this function from `dqlite/src/server.c:524 // static void stopCb(uv_async_t *stop)`.